backend/src/apiserver/resource/resource_manager.go

```diff
@@ -378,6 +385,19 @@ func (r *ResourceManager) CreateRun(apiRun *api.Run) (*model.RunDetail, error) {
 		return nil, err
 	}
 
+	// Check if the number of workflow CRDs is already over the specified threshold. If yes, delete the oldest workflow in the persisted final state.
+	// The cleanup of excessive workflows here is just a best-effort attempt, since we have Persistent Agent to do garbage collection.
+	workflows, err := r.getWorkflowClient(namespace).List(v1.ListOptions{})
```

Comment:
I don't think this can scale, getting total number of workflows is a very expensive api request when number of workflows is large.
e.g. when there are 7000 workflows, it took about 5 minutes to get the total number of workflows by `kubectl get workflows`.

Reply:
That's way we have to set the max flag to a scalable number to avoid any list call (either here or in PA) ever needing to list an enormous list. In other words, we should never allow the list to grow big in the first place.

Reply:
Previously, the cluster worked properly with 7000 workflows.
Introducing a change that helps users notice there are too many workflows is a great feature I agree, but sacrificing scalability so much doesn't seem ideal to me.

Shall we add some local caching or rough estimation mechanism? This feature doesn't need an accurate workflow count to work properly.

Reply:
an early idea: can we use `workflow watch` api instead?

## Design
api server can watch workflow resource updates and maintain a total workflow count using a daemon.
so that each KFP api call only need to query the cached workflow count. the cached count may not be 100% accurate right now (it could have a little delay), but that's enough for a feature to gate maximum number of workflows.

Reply:
not sure if there are up-to-date recommendations from kubernetes community, but this seems to match design in https://kubernetes.io/blog/2015/09/kubernetes-performance-measurements-and/

> Rewriting controllers to be watch-based: Previously they were relisting objects of a given type every few seconds, which generated a huge load on the apiserver.